src/main/java/com/uid2/shared/store/ScopedStoreReader.java

```diff
@@ -0,0 +1,69 @@
+package com.uid2.shared.store;
+
+import com.uid2.shared.Utils;
+import com.uid2.shared.attest.UidCoreClient;
+import com.uid2.shared.cloud.CloudStorageException;
+import com.uid2.shared.cloud.ICloudStorage;
+import com.uid2.shared.store.parser.ParsingResult;
+import com.uid2.shared.store.parser.Parser;
+import com.uid2.shared.store.scope.StoreScope;
+import io.vertx.core.json.JsonObject;
+import io.vertx.core.logging.Logger;
+import io.vertx.core.logging.LoggerFactory;
+
+import java.io.InputStream;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicReference;
+
+public class ScopedStoreReader<T> {
+    private static final Logger LOGGER = LoggerFactory.getLogger(ScopedStoreReader.class);
+    private final ICloudStorage metadataStreamProvider;
+    private final StoreScope scope;
+    private final Parser<T> parser;
+    private final String dataTypeName;
+    private final ICloudStorage contentStreamProvider;
+    private final AtomicReference<T> latestSnapshot = new AtomicReference<T>();
+
+    public ScopedStoreReader(ICloudStorage fileStreamProvider, StoreScope scope, Parser<T> parser, String dataTypeName) {
+        this.metadataStreamProvider = fileStreamProvider;
+        this.scope = scope;
+        this.parser = parser;
+        this.dataTypeName = dataTypeName;
+        if (fileStreamProvider instanceof UidCoreClient) {
+            this.contentStreamProvider = ((UidCoreClient) fileStreamProvider).getContentStorage();
+        } else {
+            this.contentStreamProvider = fileStreamProvider;
+        }
+    }
+
+    public CloudPath getMetadataPath() { return scope.getMetadataPath(); }
+    public T getSnapshot() {
+        return latestSnapshot.get();
+    }
+
+    public JsonObject getMetadata() throws Exception {
+        String cloudPath = getMetadataPath().toString();
+        if (isFileMissing(cloudPath)) {
+            return null;
+        }
+        return Utils.toJsonObject(this.metadataStreamProvider.download(cloudPath));
+    }
+
+    private boolean isFileMissing(String cloudPath) throws CloudStorageException {
+        List<String> files = metadataStreamProvider.list(cloudPath);
+        return files.isEmpty();
+    }
+    private long loadContent(String path) throws Exception {
+        final InputStream inputStream = this.contentStreamProvider.download(path);
+        final ParsingResult<T> parsed = parser.deserialize(inputStream);
+        this.latestSnapshot.set(parsed.getData());
+        LOGGER.info(String.format("Loaded %d %s", parsed.getCount(), dataTypeName));
+        return parsed.getCount();
+    }
+
+    public long loadContent(JsonObject metadata, String dataType) throws Exception {
```

Comment:
Would you need to do `null` checks for all the `loadContent()` that consumes the `JsonObject metadata` directly? Since `getMetadata()` may potentially return null

Reply:
Good point, a good question what we should do if it is null (ie there is no metadata). Probably the appropriate thing is to throw a better exception than a null pointer exception since if we get null from getMetadata we shouldn't really loadContent and should create it instead.

Reply:
If we go with the approach of creating when it's null, I guess we don't need to throw an exception right?

Reply:
It should happen but nothing guarantees that this is how it will go. Exceptions are exactly for that - when things go not in the way you expect them to go. Additionally only Admin would do the writing - if another service encounters this situation, if eg Core got pointed to an empty directory, they can't do more than error.

Reply:
Ok resolved and added testing around the whole area.